File: Exercism Codes/C#/Satellite.cs
public record Tree
{
    public char Value { get; set; }
    public Tree? Left { get; set; }
    public Tree? Right { get; set; }

    public Tree(char value, Tree? left = null, Tree? right = null)
    {
        Value = value;
        Left = left;
        Right = right;
    }
}

public static class Satellite
{
    public static Tree? TreeFromTraversals(char[] preOrder, char[] inOrder)
    {
        HashSet<char> inOrderSet = new HashSet<char>(inOrder);
        HashSet<char> preOrderSet = new HashSet<char>(preOrder);

        if (inOrderSet.Count != inOrder.Length || preOrderSet.Count != preOrder.Length)
        {
            throw new ArgumentException("traversals must contain unique items");
        }
        if (preOrder.Length != inOrder.Length)
        {
            throw new ArgumentException("traversals must have the same length");
        }
        if (!preOrderSet.SetEquals(inOrderSet))
        {
            throw new ArgumentException("traversals must have the same elements");
        }

        int preOrderIndex = 0;

        Tree? BuildTree(char[] preOrder, char[] inOrder, int start, int end)
        {
            if (start > end)
            {
                return null;
            }

            Tree n = new Tree(preOrder[preOrderIndex++], null, null);

            if (start == end)
            {
                return n;
            }

            int inOrderIndex = Search(inOrder, start, end, n.Value);
            n.Left = BuildTree(preOrder, inOrder, start, inOrderIndex - 1);
            n.Right = BuildTree(preOrder, inOrder, inOrderIndex + 1, end);
            return n;
        }

        return BuildTree(preOrder, inOrder, 0, inOrder.Length - 1);
    }

    private static int Search(char[] inOrder, int start, int end, char n)
    {
        for (int i = start; i <= end; i++)
        {
            if (inOrder[i] == n)
            {
                return i;
            }
        }
        return -1;
    }
}

File: Exercism Codes/JavaScript/satellite.js
//
// This is only a SKELETON file for the 'Satellite' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const treeFromTraversals = (preorder, inorder) => {
  const preOrderSet = new Set(preorder);
  const inOrderSet = new Set(inorder);
  if (preorder.length != preOrderSet.size || inorder.length != inOrderSet.size) {
    throw new Error("traversals must contain unique items");
  }
  if (preorder.length != inorder.length) {
    throw new Error("traversals must have the same length");
  }
  if (preOrderSet.symmetricDifference(inOrderSet).size !== 0) {
    throw new Error("traversals must have the same elements");
  }
  let preOrderIndex = 0;

  const buildTree = (preorder, inorder, start, end) => {
    if (start > end) {
      return {};
    }
    const n = { value: preorder[preOrderIndex++], left: {}, right: {} };
    if (start == end) {
      return n;
    }
    const inOrderIndex = search(inorder, start, end, n.value);
    n.left = buildTree(preorder, inorder, start, inOrderIndex - 1);
    n.right = buildTree(preorder, inorder, inOrderIndex + 1, end);

    return n;
  };

  return buildTree(preorder, inorder, 0, inorder.length - 1);
};

const search = (inorder, start, end, n) => {
  for (let i = start; i <= end; i++) {
    if (inorder[i] == n) {
      return i;
    }
  }
  return -1;
};
